### PR TITLE
OOP school library: decorate a class

### PR DESCRIPTION
**Pull request by jheart-vic**
#### Interface
- Created a class `Nameable`.
   - Implemented  a method called `correct_name` that will raise a `NotImplementedError`.

#### Turned my Person class to Nameable
- Made sure that your Person class inherits from Nameable
- Made  sure that this class has a method `correct_name` implemented. It should simply return the name attribute.

#### Prepared base Decorator
- Made sure that it inherits from Nameable.
- In the constructor assign a nameable object from params to an instance variable.
- Implemented the `correct_name` method that returns the result of the correct_name method of the `@nameable`.

#### Prepared CapitalizeDecorator and TrimmerDecorator
- For the CapitalizeDecorator:
    - Created a class that inherits from the base Decorator class.
    - Implemented a method `correct_name` that capitalizes the output of `@nameable.correct_name`.
- For the TrimmerDecorator:
    - Createda class that inherits from the base Decorator class.
    - Implemented a method `correct_name` that makes sure that the output of `@nameable.correct_name` has a maximum of 10 characters. If it's longer it should trim the word.

### CODE REVIEW

File: person.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  end
+
+  def of_age?
+    @age >= 18
+  end
+
+  def can_use_services?
+    of_age? || @parent_permission
+  end
+
```
Comment (ShahierNashaat) [inline person.rb:18]: - [ ] Kindly be noticed that the `Person` class should have a method `correct_name`.
```ruby
def correct_name
   name
end
```

File: person.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  private :of_age
+  public :can_use_services
```
Comment (ShahierNashaat) [inline person.rb]: - [ ] Please write the names of the methods right.
```ruby
private :of_age?
public :can_use_services?
```

File: person.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+end
+
+person = Person.new(22, `maximilianus`)
```
Comment (ShahierNashaat) [inline person.rb]: - [ ] Please use a single quote `'` when you write string.
```ruby
person = Person.new(22, 'maximilianus')
```

File: person.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
+class Person < Nameable
```
Comment (ShahierNashaat) [inline person.rb:3]: - [ ] Please require the `nameable.rb` file at the top so you can use it.
```ruby
require './nameable'

class Person < Nameable
```

File: nameable.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  end
+end
+
+class Decorator < Nameable
+  def initialize(nameable)
+    super()
+    @nameable = nameable
+    Nameable.correct_name(@nameable)
+  end
+end
```
Comment (ShahierNashaat) [inline nameable.rb]: - [ ] Please implement`correct_name` method in the `Decorator` class.
```ruby
class Decorator < Nameable
    attr_accessor :nameable
  
    def initialize(nameable)
      @nameable = nameable
    end
  

    def correct_name
      @nameable.correct_name
    end
end
```

File: nameable.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+class CapitalizeDecorator < Decorator
+  def correct_name
+    @nameable.capitalize
+  end
```
Comment (ShahierNashaat) [inline nameable.rb]: - [ ] Kindly notice that the `correct_name` method in the `CapitalizeDecorator` should be implemented like this.
```ruby
def correct_name
  @nameable.correct_name.capitalize
end
```

File: nameable.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+end
+
+class TrimmerDecorator < Decorator
+  def correct_name
+    @nameable.strip
+  end
```
Comment (ShahierNashaat) [inline nameable.rb]: - [ ] Kindly notice that the `correct_name` method of the `TrimmerDecorator` should be implemented like this.
```ruby
def correct_name
  @nameable.correct_name.length > 10 ? @nameable.correct_name[0..9] : @nameable.correct_name
end
```

**Review by ShahierNashaat — CHANGES_REQUESTED** [review]
Hi @jheart-vic, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight 💯 

- Correct use of GitFlow ✔️
- Descriptive commit messages ✔️
- Professional looking pull request ✔️
- Your readme looks professional ✔️
- Your code is clean and well documented ✔️

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: README.md [hunk 8]
```diff
@@ -1,60 +1,45 @@
...
 
-- Major languages
-- Frameworks
-- Technologies used
+- RUBY
 
-## Live Demo (if available)
+## Live Demo (Not available)
 
 [Live Demo Link](https://livedemo.com)
```
Comment (danmainah) [inline README.md:11]: - [ ] [optional It's not advisable to have a demo link in the readme if you don't have a demo link.

**Review by danmainah — APPROVED** [review]
Hi @jheart-vic , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🎉 🎉 

## Highlights
✔️ No linter errors
✔️ Gitflow was correctly used

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**